src/IROperator.cpp

```diff
@@ -1533,7 +1533,7 @@ Expr promise_clamped(const Expr &value, const Expr &min, const Expr &max) {
     return Internal::Call::make(value.type(),
```

Comment:
The meaning of promise_clamped is location-sensitive. You can't safely move that piece of IR around, so it's not pure.

Reply:
Reverting, but I'm having a hard time envisioning an example where that would be the case. Can you provide an example for posterity here?

Reply:
```
for (int x = 0; x < 10; x++) {
  if (2*x < 10) {
    f(promise_clamped(x, 0, 4));
  }
}
```

Reply:
In the example above, I'm not sure if it's safe to move the x outside the if

Then again... hrm. The only users should be inside the if, so how could it matter...

Reply:
I guess the implication is that even evaluating a promise clamped, where the arg is not actually clamped, is undefined.

Reply:
@dsharletg Maybe promise_clamped is actually pure? What do you think?

Reply:
I think it is pure technically, but I think maybe you are thinking of uses of promise_clamped that would break if it were moved (to a let)? So while it is in fact pure, we mark it not pure to avoid this problem?

Reply:
By uses, I mean consumers of promise_clamped's information, not the code that generates promise_clamped.